### PR TITLE
some modernization of the memory handing in the canonicalization code

### PR DESCRIPTION
**Pull request by greglandrum**
Switch to using unique_ptrs instead of naked pointers and re-arrange the allocation a bit.

I benchmarked this and there's no significant performance impact.

### CODE REVIEW

File: Code/GraphMol/new_canon.cpp [hunk 1]
```diff
@@ -160,20 +165,13 @@ void compareRingAtomsConcerningNumNeighbors(Canon::canon_atom *atoms,
     unsigned currentRNIdx = 0;
     atoms[idx].neighborNum.reserve(1000);
     atoms[idx].revistedNeighbors.assign(1000, 0);
-    char *visited = (char *)malloc(nAtoms * sizeof(char));
-    CHECK_INVARIANT(visited, "allocation failed");
-    memset(visited, 0, nAtoms * sizeof(char));
+
+    memset(visited.get(), 0, nAtoms * sizeof(char));
```
Comment (bp-kelley) [inline Code/GraphMol/new_canon.cpp:169]: Nice optimization!  Does the lack of calling malloc in the loop speed this up a bunch?

File: Code/GraphMol/new_canon.cpp [hunk 2]
```diff
@@ -661,10 +652,7 @@ void initChiralCanonAtoms(const ROMol &mol,
 
 void freeCanonAtoms(std::vector<Canon::canon_atom> &atoms) {
   for (auto &atom : atoms) {
-    if (atom.nbrIds) {
-      free(atom.nbrIds);
-      atom.nbrIds = nullptr;
-    }
+    atom.nbrIds.reset();
```
Comment (bp-kelley) [inline Code/GraphMol/new_canon.cpp:655]: Are the atoms reused?  You might not need this with the unique_ptr

**Review by bp-kelley — APPROVED** [review]
LGTM - might have an unnecessary unique_ptr.reset()

File: Code/GraphMol/new_canon.cpp (hunk 1 above)
Comment (greglandrum) [reply Code/GraphMol/new_canon.cpp:169]: nope, there was no significant performance impact of the changes in this PR

File: Code/GraphMol/new_canon.cpp (hunk 2 above)
Comment (greglandrum) [reply Code/GraphMol/new_canon.cpp:655]: good point

**Pull request merged**